src/metabase/analyze/fingerprint/fingerprinters.clj

```diff
@@ -63,25 +66,37 @@
                                       ~v
                                       (catch Throwable _#))]))))
 
-(defmacro ^:private with-reduced-error
-  [msg & body]
-  `(let [result# (sync-util/with-error-handling ~msg ~@body)]
-     (if (instance? Throwable result#)
-       (reduced result#)
-       result#)))
-
 (defn with-error-handling
   "Wrap `rf` in an error-catching transducer."
   [rf msg]
+  ;; This function is written in a specific way to ensure that try-catch blocks don't produce closures.
   (fn
-    ([] (with-reduced-error msg (rf)))
+    ([]
+     (if sync-util/*log-exceptions-and-continue?*
+       (try (rf)
+            (catch Throwable e
+              (metabase.util.log/warn e msg)
+              (reduced e)))
+       (rf)))
     ([acc]
-     (unreduced
-      (if (or (reduced? acc)
-              (instance? Throwable acc))
-        acc
-        (with-reduced-error msg (rf acc)))))
-    ([acc e] (with-reduced-error msg (rf acc e)))))
+     (cond (or (reduced? acc)
+               (instance? Throwable acc))
+           (unreduced acc)
+
+           sync-util/*log-exceptions-and-continue?*
+           (try (unreduced (rf acc))
+                (catch Throwable e
+                  (metabase.util.log/warn e msg)
+                  e))
+
+           :else (rf acc)))
+    ([acc e]
+     (if sync-util/*log-exceptions-and-continue?*
+       (try (rf acc e)
+            (catch Throwable e
+              (metabase.util.log/warn e msg)
+              (reduced e)))
+       (rf acc e)))))
```

Comment:
I can make it a private macro as long as it will be really private and won't have to generalize, because it's a pain to make it generalizable and have it make sense for an outer user.